IPIP/0000-double-hash-dht.md

```diff
@@ -0,0 +1,337 @@
+- **`multiaddrs`** are the [network addresses](https://github.com/libp2p/specs/tree/master/addressing) associated with a `PeerID`. It represents the location(s) of the peer.
+- **`KeyPrefix`** is defined as a prefix of length `l` bits of `HASH2`. `KeyPrefix` is represented by a `byte` concatenated with a variable sized array of bytes, containing at most 32 bytes. The leading `byte` represents the binary representation of `l - 1`, making prefixes of length `256` possible, but not prefixes of length `0`. The trailing byte array is of length `ceil(l/8)` bytes, and its content is the bits prefix right padded with zeros.
+- **`ServerKey`** is defined as `SHA256(bytes("CR_SERVERKEY") || MH)`. It is derived from the `MH`. The Content Provider communicates `ServerKey` to the DHT Servers during the Publish Process. The DHT Servers use `ServerKey` to encrypt `TS`, `Signature` and Content Providers `multiaddrs` sent to the Client when some Provider Records match the requested `Prefix`. `ServerKey` is represented as a 32-byte array.
+- **`TS`** is the [Unix Timestamp](https://en.wikipedia.org/wiki/Unix_time) corresponding content publish time. `TS` is represented as a 32-bit **unsigned** Integer, allowing timestamps to range from `1970-01-01T00:00:00Z` to `2106-02-07T06:28:15Z` before reaching the overflow.
+- **`CPPeerID`** is the `PeerID` of the Content Provider for a specific `CID`.
+- **`EncPeerID`** is the result of the encryption of `CPPeerID` using `MH` as encryption key and a random nonce `AESGCM(MH, Nonce, CPPeerID)`. `EncPeerID` contains the [varint](https://github.com/multiformats/multicodec/pull/314) of the encryption algorithm used (AES), the bytes array of the encrypted payload, and the `Nonce`. `Nonce` is a randomly generated 12-byte array. The format of `EncPeerID` is [`0xa501`, `Nonce`, `payload_len`, `AESGCM(MH, Nonce, CPPeerID)`].
+- **`Signature`** is the signature of the `EncPeerID` encrypted payload (not including the varint nor the nonce) and `TS` using the Content Provider's private key, either with ed25519 or rsa signature algorithm, depending on the keys of the Content Provider.
+- **Provider Record** is defined as a pointer to the storage location of some content identified by `CID` or `HASH2`. A Provider Record consists on the following fields: [`EncPeerID`, `TS`, `Signature`].
+- **Provider Store** is the data structure on the DHT Servers used to store the Provider Records. Its structure is a nested dictionary/map: `HASH2` -> `ServerKey` -> [`CPPeerID`, `EncPeerID`, `TS`, `Signature`]. There is only one single correct `ServerKey` for each `HASH2`. However, any peer can forge a valid Publish request (with invalid `EncPeerID` but valid `Signature`) undetected by the DHT Server. The DHT server isn't able to distinguish which `ServerKey` is correct as it doesn't have the knowledge of `MH`, hence it has to keep both and serve both upon request for `HASH2`.
+
+### Current DHT
```

Comment:
As we want to replace the current DHT with the double hash one, I don't think it makes sense to create a spec now for the dht to be deprecated.

However, I am happy to contribute to a new dht spec describing the overall behavior of the dht reusing and maybe enhancing https://github.com/libp2p/specs/tree/master/kad-dht. This new dht spec can be merged when the double hash dht goes live, so that we don't launch any undocumented feature.

Reply:
I agree it would be worthwhile to get the current state documented and apply deltas on top.  

@guillaumemichel : I hear your concern about not wanting to do throwaway work (i.e., document something that is going to be changed by this IPIP)?  That said, how much throwaway work is it?  Also, I think it would be nice for your PR to only show the changes from the past vs. also bundling in all the other aspects of the DHT.

Reply:
I think that most of the spec can be reused, and the deltas are already described in this PR and in [Notion](https://pl-strflt.notion.site/Double-Hashing-for-Privacy-ff44e3156ce040579289996fec9af609). I'll add the legacy dht spec to my backlog.

Currently, this PR only focuses on the changes, and also describes the concepts required to understand the change.